DOCS-8034-vm-reorganization-kt (#1038)
mulesoft/docs-connectors: vm/2.0/modules/ROOT/pages/index.adoc

Support Category: https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Select]

> Anypoint Connector for Virtual Machine (VM Connector) handles intra-app and inter-app communication through asynchronous queues that can be transient or persistent.

Reviewer: Can you think of a better word than "handles"?
I like to avoid words that have more than one meaning when possible, for translation.
"manages"? "processes"? "performs"? Whatever you think is best.

Reviewer: In addition to the word substitution, this suggestion:
```suggestion
Anypoint Connector for Virtual Machine (VM Connector) handles intra-app and inter-app communication through either transient or persistent asynchronous queues:
```

> * Transient queues are faster than persistent queues, but they are not reliable in the case of a system crash

Reviewer: ```suggestion
* Transient queues are faster than persistent queues, but they are not reliable if the system crashes.
```

Reviewer: Is this the end of all text about VM Connector working with transient queues?

> * Persistent queues are slower but reliable

Reviewer: ```suggestion
* Persistent queues are slower but reliable.
```

> When running on a single instance, persistent queues work by serializing and storing the contents on the disk. However, when running in cluster mode, the memory grid backs the persistent queues. Thereby, when a flow uses the VM Connector to publish content to a queue, Mule runtime engine determines whether to process the message in the same origin node or to send the message to the cluster to pick another node. The VM Connector facilitates the load distribution across a cluster.

Reviewer: What is VM Connector's role in the single-instance scenario?

Reviewer: What does this mean: "the memory grid _backs_ the persistent queues"?
Do we mean "backs up" or "supports"? Or do we mean that the memory grid takes on the job of the disk on a single instance?

Reviewer: So, I'm a little confused. Is this topic about how VM Connector helps with load distribution in a scenario of a persistent queue and cluster mode? Because transient queues and single-instance mode seem kind of downplayed up to this point.

Reviewer: Maybe the info would make more sense to me ordered like this (if it is technically accurate):

VM Connector facilitates communication that uses either transient and persistent data queues:
* Transient queues process data more quickly, but are more susceptible to data loss if the system crashes.
   (Text that tells why, and what roll the connector plays with transient data queues)
* Persistent queues process data more slowly but are less susceptible to data loss.
   There are two scenarios: single-instance processing and cluster-mode operation:
          ** Single
               How it works, what role connector plays
          ** Cluster
               How it works, what role connector plays
     In either case, data in persistent queues must be serializable. Although you can use Kyro to meet this
     requirement, it is limited <in this way>. It is better <to not use Kyro, or to use Kyro with these additional
     requirements>:
         ** Simple
         ** Serializable interface....

Reviewer: OR
Maybe we should have separate subtopics rather than two bullets, because the persistent data queue scenario is so much more complicated than the transient:

...transient or persistent data queues

==Communication Using a Transient Data Queue
as above

==Communication Using a Persistent Data Queue
as above

Reviewer: Questions about existing text:
```suggestion
When running on a single instance, persistent queues work by serializing and storing the contents on the disk. However, when running in cluster mode, the memory grid backs the persistent queues. Thereby, when a flow uses the VM Connector to publish content to a queue, Mule runtime engine determines whether to process the message in the same origin node or to send the message to the cluster to pick another node. The VM Connector facilitates the load distribution across a cluster.
```
I don't understand the chronology. This text makes it seem...:
1. How single instance processing works
2. How cluster mode processing works
3. How Mule decides which whether to use single instance (same node) or cluster mode (distributed)???
4. VM Connector helps only with cluster mode?

Reviewer: I can't really edit this above text, because the order of information is not clear to me.
Maybe another connector peer writer can help, using my questions?

> In either case, when using persistent queues the contents you send must be serializable. Even though you can enable Kryo serialization to get a wider range of serializable values, Kryo does have some limitations. Therefore, try to:

Reviewer: ```suggestion
In either single-instance or cluster-mode processing, when using persistent queues, the data you send must be serializable. Although you can enable Kryo serialization to get a wider range of serializable values, Kryo does have some limitations. Therefore, try to:
```
Question about "Therefore, try to:"
Are we saying that you must try to do these things to address the Kyro limitations, or are we saying that you should try to do these things INSTEAD of using Kyro?
And do we need to specify what those limitations are?

Reviewer: GLOBALLY review how this topic variously refers to the information in the communication: contents? data? values? message?
Can we standardize to fewer or just one term?

> * Keep your values simple because complex structures might result in serialization errors or performance issues if they are not easy to serialize

Reviewer: ```suggestion
* Keep your values simple to avoid errors or performance issues when serializing complex structures.
```

> * Make sure complex Java object implements the `Serializable` interface and that they conform to the Java Bean contract

Reviewer: What is the subject of this sentence: object or objects? My guess...:
```suggestion
* Ensure that complex Java objects implement the `Serializable` interface and that they conform to the Java Bean contract.
```
...but does an object "implement" an interface?

> * Use streams, JSON objects, maps, and so on as long as the associated values comply with the previous recommendations

Reviewer: ```suggestion
* Use streams, JSON objects, maps, and so on only if the associated values comply with the previous recommendations.
```

== Prerequisites

To use this connector, you must be familiar with:

* Anypoint Connectors
* Mule runtime engine (Mule)
* Elements and global elements in a Mule flow
* Creating a Mule app using Anypoint Studio (Studio)

Before creating an app, you must have access to Anypoint Studio.

== Common Use Cases for the Connector

> Use the VM Connector when you want to:

Reviewer: GLOBALLY remove "the" from before "VM Connector" unless you follow the product name with a common noun.
See style ref. under "connector naming."
```suggestion
Use VM Connector when you want to:
```

* Pass messages from one flow to another through a queuing mechanism, instead of using a `<flow-ref />` directly
* Distribute work across a cluster
* Communicate with different apps that are running in the same Mule domain

> * Perform simple queueing that does not justify the usage of a full JMS broker

Reviewer: I wonder if the inclusion of this list does not remove the need for all the intro text about different types of queues and how they work...?

Reviewer: ```suggestion
* Pass messages from one flow to another through a queuing mechanism, instead of using `<flow-ref />` directly
* Distribute work across a cluster
* Communicate with different apps that are running in the same Mule domain
* Perform simple queueing that does not justify the use of a full JMS broker
```

Support Category: https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Select]

Anypoint Connector for Virtual Machine (VM Connector) manages intra-app and inter-app communication through either transient or persistent asynchronous queues:

* Transient queues are faster than persistent queues, but they are not reliable if the system crashes.
* Persistent queues are slower but reliable.

== Prerequisites

To use this connector, you must be familiar with:

* Anypoint Connectors
* Mule runtime engine (Mule)
* Elements and global elements in a Mule flow
* Creating a Mule app using Anypoint Studio (Studio)

Before creating an app, you must have access to Anypoint Studio.

== Common Use Cases for the Connector

Use VM Connector when you want to:

* Pass messages from one flow to another through a queuing mechanism, instead of using a `<flow-ref />` directly
* Distribute work across a cluster
* Communicate with different apps that are running in the same Mule domain
* Perform simple queueing that does not justify the use of a full JMS broker

For examples, see xref:vm-examples.adoc[Examples].

== Limitations for Persistent Queues

> When running a Mule application on a single runtime instance mode, persistent queues work by serializing and storing the contents on the disk. However, when running the Mule application in cluster runtime instance mode, the persistent queues are backed up in the memory grid. Thereby, when a flow uses VM Connector to publish content to a queue, Mule runtime engine determines whether to process the message in the same origin node or to send the message to the cluster to pick another node.Using VM Connector facilitates the load distribution across a cluster. +

Reviewer 2: ```suggestion
When running a Mule application in single runtime instance mode, persistent queues work by serializing and storing the contents on the disk. However, when running the Mule application in cluster runtime instance mode, the persistent queues are backed up in the memory grid. Thereby, when a flow uses VM Connector to publish content to a queue, Mule runtime engine determines whether to process the message in the same origin node or to send the message to the cluster to pick another node. Using VM Connector facilitates the load distribution across a cluster. +
```

In either single runtime instance mode or cluster runtime instance mode, when using persistent queues, the data you send must be serializable. Although you can enable Kryo serialization to get a wider range of serializable values, Kryo does have some limitations. Therefore, when using persistent queues try to:

* Keep your values simple because complex structures might result in serialization errors or performance issues if they are not easy to serialize.

> * Ensure complex Java objects implement the `Serializable` interface and that they conform to the Java Bean contract.

Reviewer 2: ```suggestion
* Ensure complex Java objects implement the `Serializable` interface and that they conform to the JavaBean contract.
```
